common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -162,6 +162,11 @@ public ConscryptSession provideSession() {
             }
         }));
 
+    /**
+     * The handshake session object exposed externally from this class.
+     */
+    private SSLSession externalHandshakeSession;
```

Comment:
I don't like this specific part of the change, it spreads out logic that only matters to getHandshakeSession() across three different places in the class.  I prefer the previous version of getHandshakeSession(), since it's really obvious exactly what that method does and where the value comes from.

Reply:
The reason for this change is so that we only create the wrapper `ExtendedSSLSession` once. Otherwise, we'll create a new one every time a call is made to `getHanshakeSession`. Maybe it's not a big deal. WDYT?

Reply:
@flooey also, if we revert this then `getHandshakeSession` can suffer from the same problem that `getSession` did WRT the caller holding onto a reference that becomes invalid.

Reply:
@nmittler I think it's fine to just return externalSession from getHandshakeSession() when the state says we shouldn't return null.  Then we don't have to make any extra wrappers.  The Javadoc implies that holding onto the session after the handshake completes should work, so I think that's fine ("For example, if a remote server will be sending a Certificate chain, but that chain has yet not been processed, the getPeerCertificates method of SSLSession will throw a SSLPeerUnverifiedException. Once that chain has been processed, getPeerCertificates will return the proper value.").

Reply:
@flooey returning `externalSession` from `getHandshakeSession` won't work because the external session will be a `NullSession` until the handshake is completed.  The purpose of `getHandshakeSession`, however, is to provide the active session during the handshake.

Reply:
@nmittler Oh, good point.  In that case, I think creating the wrapper on demand is fine.  I think basically nobody calls getHandshakeSession(), and an even smaller portion call it more than once, so on average we won't ever have to create the wrapper this way.